tests/proptest.rs

```diff
@@ -0,0 +1,436 @@
+                                Range::higher_than(s[c].0)
+                            } else if c == d {
+                                Range::exact(s[c].0)
+                            } else {
+                                Range::between(s[c].0, s[d].0 + 1)
+                            },
+                        ))
+                    }
+                }
+
+                let mut solver = OfflineDependencyProvider::<N, NumberVersion>::new();
+
+                let complicated_len = std::cmp::min(complicated_len, list_of_pkgid.len());
+                let complicated: Vec<_> = if reverse_alphabetical {
+                    &list_of_pkgid[..complicated_len]
+                } else {
+                    &list_of_pkgid[(list_of_pkgid.len() - complicated_len)..]
+                }
+                .iter()
+                .map(|(x, _)| (x.0.clone(), x.1))
+                .collect();
+
+                for ((name, ver), deps) in list_of_pkgid {
+                    solver.add_dependencies(
+                        name,
+                        ver,
+                        deps.unwrap_or_else(|| vec![(bad_name.clone(), Range::any())]),
+                    );
+                }
+
+                (solver, complicated)
+            },
+        )
+}
+
+/// This test is to test the generator to ensure
+/// that it makes registries with large dependency trees
+#[test]
+fn meta_test_deep_trees_from_strategy() {
+    use proptest::strategy::ValueTree;
+    use proptest::test_runner::TestRunner;
+
+    let mut dis = [0; 21];
+
+    let strategy = registry_strategy(0u16..665, 666);
+    let mut test_runner = TestRunner::deterministic();
+    for _ in 0..128 {
+        let (dependency_provider, cases) = strategy
+            .new_tree(&mut TestRunner::new_with_rng(
+                Default::default(),
+                test_runner.new_rng(),
+            ))
+            .unwrap()
+            .current();
+
+        for (name, ver) in cases {
+            let res = resolve(&dependency_provider, name, ver);
+            dis[res
+                .as_ref()
+                .map(|x| std::cmp::min(x.len(), dis.len()) - 1)
+                .unwrap_or(0)] += 1;
+            if dis.iter().all(|&x| x > 0) {
+                return;
+            }
+        }
+    }
+
+    panic!(
+        "In {} tries we did not see a wide enough distribution of dependency trees! dis: {:?}",
+        dis.iter().sum::<i32>(),
+        dis
+    );
+}
+
+proptest! {
+    #![proptest_config(ProptestConfig {
+    max_shrink_iters:
+        if std::env::var("CI").is_ok() {
+            // This attempts to make sure that CI will fail fast,
+            0
+        } else {
+            // but that local builds will give a small clear test case.
+            2048 // u32::MAX
+        },
+        result_cache: prop::test_runner::basic_result_cache,
+        .. ProptestConfig::default()
+    })]
+
+    #[test]
+    /// This test is mostly for profiling
```

Comment:
Because it uses strings for package ids right?

Reply:
Yes. It was one of the first tests I wrote after getting `registry_strategy` working. It may no longer be pulling its weight. I wanted to compare a profile with expensive clones vs one with cheap clones. The other use the faster version ids to not waste CI time, but I want something to test that we are not cloning way way to much. I am open to suggestions.

Reply:
It's good to still have a strings example in my opinion, to be closer to real usage when benchmarking things that depend on hash maps. If it's not needed on CI though, is there a way to deactivate a bunch of tests specifically when on CI?

Reply:
There is always `if std::env::var("CI").is_ok() {return;}`. If it is still worth running then I am ok running on CI. When CI is a bottleneck we can revisit.